### PR TITLE
channel mode +M

### PR DESCRIPTION
**Pull request by empus**
New channel mode to prevent unauthenticated users from speaking or using part messages in a channel (unless opped or voiced).

Helpful to combat channel abuse (especially the recent troubles) where +r is too restrictive and the use of +Dm with bots is too complex for many users.  This new mode could be used either permanently for many channels, or temporarily during periods of abuse.

### CODE REVIEW

File: include/channel.h [hunk 1]
```diff
@@ -113,12 +113,15 @@ struct Client;
 #define MODE_APASS	0x200000
 #define MODE_WASDELJOINS 0x400000 	/**< Not DELJOINS, but some joins
 					 * pending */
+
+#define MODE_MODERATENOREG 0x2000000    /**< +M Moderate unauthed users */
```
Comment (entrope) [inline include/channel.h]: Other channel modes are not surrounded by blank lines; this one should not be, either.

File: ircd/channel.c [hunk 2]
```diff
@@ -708,6 +708,10 @@ int member_can_send_to_channel(struct Membership* member, int reveal)
   if (member->channel->mode.mode & MODE_MODERATED)
     return 0;
 
+  /* If only logged in users may join and you're not one, you can't speak. */
```
Comment (entrope) [inline ircd/channel.c]: This duplicates too much of the following lines (715-718).

File: ircd/channel.c [hunk 3]
```diff
@@ -1966,8 +1973,8 @@ modebuf_mode(struct ModeBuf *mbuf, unsigned int mode)
 
   mode &= (MODE_ADD | MODE_DEL | MODE_PRIVATE | MODE_SECRET | MODE_MODERATED |
 	   MODE_TOPICLIMIT | MODE_INVITEONLY | MODE_NOPRIVMSGS | MODE_REGONLY |
-           MODE_NOCOLOR | MODE_NOCTCP |
-           MODE_DELJOINS | MODE_WASDELJOINS | MODE_REGISTERED);
+           MODE_NOCOLOR | MODE_NOCTCP | MODE_MODERATENOREG | MODE_DELJOINS |
```
Comment (entrope) [inline ircd/channel.c]: This should not move MODE_DELJOINS.
```suggestion
           MODE_DELJOINS | MODE_NOCOLOR | MODE_NOCTCP | MODE_MODERATENOREG |
```

File: include/channel.h (hunk 1 above)
Comment (empus) [reply include/channel.h]: FIxed

File: ircd/channel.c [hunk 4]
```diff
@@ -1966,8 +1969,8 @@ modebuf_mode(struct ModeBuf *mbuf, unsigned int mode)
 
   mode &= (MODE_ADD | MODE_DEL | MODE_PRIVATE | MODE_SECRET | MODE_MODERATED |
 	   MODE_TOPICLIMIT | MODE_INVITEONLY | MODE_NOPRIVMSGS | MODE_REGONLY |
-           MODE_NOCOLOR | MODE_NOCTCP |
-           MODE_DELJOINS | MODE_WASDELJOINS | MODE_REGISTERED);
+           MODE_DELJOINS | MODE_NOCOLOR | MODE_NOCTCP | MODE_MODERATENOREG |
```
Comment (entrope) [inline ircd/channel.c]: This still moves MODE_DELJOINS for no apparent reason.

Comment (empus) [reply ircd/channel.c]: Should be solved now

**Pull request merged**